Investigate a compilation failure for linux/arm64

Hello,

It is more of a question to see if you have experienced something this. It has been noted that this specific package has started to cause compilation failures, and so far, it has always been in the context of building a Docker image for the linux/arm64 platform, which as when you build with Apple silicon or in GitHub Actions with an emulator.

The exact same workflow with no change in between was succeeded a month or so ago. It could then be something about the runner images, which periodically received updates. Is it plausible? Do you have any other guesses?

Thank you!

https://github.com/actions/runner-images/blob/ubuntu22/20230821.1/images/linux/Ubuntu2204-Readme.md
https://github.com/actions/runner-images/blob/ubuntu22/20230814.1/images/linux/Ubuntu2204-Readme.md
https://github.com/actions/runner-images/blob/ubuntu22/20230806.1/images/linux/Ubuntu2204-Readme.md
…

```
#44 149.0   Building wheel for openstep-plist (pyproject.toml): finished with status 'error'
#44 149.4   error: subprocess-exited-with-error
#44 149.4   
#44 149.4   × Building wheel for openstep-plist (pyproject.toml) did not run successfully.
#44 149.4   │ exit code: 1
#44 149.4   ╰─> [348 lines of output]
#44 149.4       <string>:7: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
#44 149.4       /tmp/pip-build-env-s9z2g2h1/overlay/lib/python3.9/site-packages/setuptools/config/setupcfg.py:293: _DeprecatedConfig: Deprecated config in `setup.cfg`
#44 149.4       !!
#44 149.4       
#44 149.4               ********************************************************************************
#44 149.4               The license_file parameter is deprecated, use license_files instead.
#44 149.4       
#44 149.4               By 2023-Oct-30, you need to update your project and remove deprecated calls
#44 149.4               or your builds will no longer be supported.
#44 149.4       
#44 149.4               See https://setuptools.pypa.io/en/latest/userguide/declarative_config.html for details.
#44 149.4               ********************************************************************************
#44 149.4       
#44 149.4       !!
#44 149.4         parsed = self.parsers.get(option_name, lambda x: x)(value)
#44 149.4       running bdist_wheel
#44 149.4       running build
#44 149.4       running build_py
#44 149.4       creating build
#44 149.4       creating build/lib.linux-aarch64-cpython-39
#44 149.4       creating build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/_version.py -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/__init__.py -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/__main__.py -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       running egg_info
#44 149.4       writing src/openstep_plist.egg-info/PKG-INFO
#44 149.4       writing dependency_links to src/openstep_plist.egg-info/dependency_links.txt
#44 149.4       writing top-level names to src/openstep_plist.egg-info/top_level.txt
#44 149.4       reading manifest file 'src/openstep_plist.egg-info/SOURCES.txt'
#44 149.4       reading manifest template 'MANIFEST.in'
#44 149.4       adding license file 'LICENSE'
#44 149.4       writing manifest file 'src/openstep_plist.egg-info/SOURCES.txt'
#44 149.4       copying src/openstep_plist/_test.cpp -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/_test.pyx -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/parser.cpp -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/parser.pxd -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/parser.pyx -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/util.cpp -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/util.pxd -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/util.pyx -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/writer.cpp -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/writer.pxd -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       copying src/openstep_plist/writer.pyx -> build/lib.linux-aarch64-cpython-39/openstep_plist
#44 149.4       running build_ext
#44 149.4       building 'openstep_plist.parser' extension
#44 149.4       creating build/temp.linux-aarch64-cpython-39
#44 149.4       creating build/temp.linux-aarch64-cpython-39/src
#44 149.4       creating build/temp.linux-aarch64-cpython-39/src/openstep_plist
#44 149.4       aarch64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -ffile-prefix-map=/build/python3.9-PN012d/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DCYTHON_TRACE_NOGIL=1 -I/install/include -I/usr/include/python3.9 -c src/openstep_plist/parser.cpp -o build/temp.linux-aarch64-cpython-39/src/openstep_plist/parser.o -std=c++11
#44 149.4       src/openstep_plist/parser.cpp: In function ‘PyObject* __pyx_f_14openstep_plist_6parser_parse_quoted_plist_string(__pyx_t_14openstep_plist_6parser_ParseInfo*, Py_UNICODE)’:
#44 149.4       src/openstep_plist/parser.cpp:4730:103: warning: ‘PyObject* PyUnicode_FromUnicode(const Py_UNICODE*, Py_ssize_t)’ is deprecated [-Wdeprecated-declarations]
#44 149.4        4730 |   __pyx_t_6 = PyUnicode_FromUnicode(((Py_UNICODE const *)__pyx_v_string.data()), __pyx_v_string.size()); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 231, __pyx_L1_error)
#44 149.4             |                                                                                                       ^
#44 149.4       In file included from /usr/include/python3.9/unicodeobject.h:1026,
#44 149.4                        from /usr/include/python3.9/Python.h:97,
#44 149.4                        from src/openstep_plist/parser.cpp:39:
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:551:42: note: declared here
#44 149.4         551 | Py_DEPRECATED(3.3) PyAPI_FUNC(PyObject*) PyUnicode_FromUnicode(
#44 149.4             |                                          ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/parser.cpp: In function ‘PyObject* __pyx_pf_14openstep_plist_6parser_string_to_number(PyObject*, PyObject*, int)’:
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:451:75: warning: ‘Py_ssize_t _PyUnicode_get_wstr_length(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         451 | #define PyUnicode_WSTR_LENGTH(op) _PyUnicode_get_wstr_length((PyObject*)op)
#44 149.4             |                                                                           ^
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:261:7: note: in expansion of macro ‘PyUnicode_WSTR_LENGTH’
#44 149.4         261 |       PyUnicode_WSTR_LENGTH(op) :                    \
#44 149.4             |       ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/parser.cpp:4908:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        4908 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_s);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here
#44 149.4         446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {
#44 149.4             |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:262:52: warning: ‘Py_UNICODE* PyUnicode_AsUnicode(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         262 |       ((void)PyUnicode_AsUnicode(_PyObject_CAST(op)),\
#44 149.4             |                                                    ^
#44 149.4       src/openstep_plist/parser.cpp:4908:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        4908 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_s);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
#44 149.4         580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
#44 149.4             |                                             ^~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:451:75: warning: ‘Py_ssize_t _PyUnicode_get_wstr_length(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         451 | #define PyUnicode_WSTR_LENGTH(op) _PyUnicode_get_wstr_length((PyObject*)op)
#44 149.4             |                                                                           ^
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:264:8: note: in expansion of macro ‘PyUnicode_WSTR_LENGTH’
#44 149.4         264 |        PyUnicode_WSTR_LENGTH(op)))
#44 149.4             |        ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/parser.cpp:4908:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        4908 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_s);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here
#44 149.4         446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {
#44 149.4             |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:279:45: warning: ‘Py_UNICODE* PyUnicode_AsUnicode(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         279 |       PyUnicode_AsUnicode(_PyObject_CAST(op)))
#44 149.4             |                                             ^
#44 149.4       src/openstep_plist/parser.cpp:4927:19: note: in expansion of macro ‘PyUnicode_AS_UNICODE’
#44 149.4        4927 |     __pyx_v_buf = PyUnicode_AS_UNICODE(__pyx_v_s);
#44 149.4             |                   ^~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
#44 149.4         580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
#44 149.4             |                                             ^~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/parser.cpp: In function ‘PyObject* __pyx_f_14openstep_plist_6parser_parse_unquoted_plist_string(__pyx_t_14openstep_plist_6parser_ParseInfo*, __pyx_opt_args_14openstep_plist_6parser_parse_unquoted_plist_string*)’:
#44 149.4       src/openstep_plist/parser.cpp:5596:67: warning: ‘PyObject* PyUnicode_FromUnicode(const Py_UNICODE*, Py_ssize_t)’ is deprecated [-Wdeprecated-declarations]
#44 149.4        5596 |     __pyx_t_3 = PyUnicode_FromUnicode(__pyx_v_mark, __pyx_v_length); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 318, __pyx_L1_error)
#44 149.4             |                                                                   ^
#44 149.4       In file included from /usr/include/python3.9/unicodeobject.h:1026,
#44 149.4                        from /usr/include/python3.9/Python.h:97,
#44 149.4                        from src/openstep_plist/parser.cpp:39:
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:551:42: note: declared here
#44 149.4         551 | Py_DEPRECATED(3.3) PyAPI_FUNC(PyObject*) PyUnicode_FromUnicode(
#44 149.4             |                                          ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/parser.cpp: In function ‘PyObject* __pyx_pf_14openstep_plist_6parser_2loads(PyObject*, PyObject*, PyObject*, int)’:
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:451:75: warning: ‘Py_ssize_t _PyUnicode_get_wstr_length(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         451 | #define PyUnicode_WSTR_LENGTH(op) _PyUnicode_get_wstr_length((PyObject*)op)
#44 149.4             |                                                                           ^
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:261:7: note: in expansion of macro ‘PyUnicode_WSTR_LENGTH’
#44 149.4         261 |       PyUnicode_WSTR_LENGTH(op) :                    \
#44 149.4             |       ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/parser.cpp:8410:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        8410 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_s);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here
#44 149.4         446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {
#44 149.4             |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:262:52: warning: ‘Py_UNICODE* PyUnicode_AsUnicode(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         262 |       ((void)PyUnicode_AsUnicode(_PyObject_CAST(op)),\
#44 149.4             |                                                    ^
#44 149.4       src/openstep_plist/parser.cpp:8410:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        8410 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_s);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
#44 149.4         580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
#44 149.4             |                                             ^~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:451:75: warning: ‘Py_ssize_t _PyUnicode_get_wstr_length(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         451 | #define PyUnicode_WSTR_LENGTH(op) _PyUnicode_get_wstr_length((PyObject*)op)
#44 149.4             |                                                                           ^
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:264:8: note: in expansion of macro ‘PyUnicode_WSTR_LENGTH’
#44 149.4         264 |        PyUnicode_WSTR_LENGTH(op)))
#44 149.4             |        ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/parser.cpp:8410:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        8410 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_s);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here
#44 149.4         446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {
#44 149.4             |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:279:45: warning: ‘Py_UNICODE* PyUnicode_AsUnicode(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         279 |       PyUnicode_AsUnicode(_PyObject_CAST(op)))
#44 149.4             |                                             ^
#44 149.4       src/openstep_plist/parser.cpp:8419:17: note: in expansion of macro ‘PyUnicode_AS_UNICODE’
#44 149.4        8419 |   __pyx_v_buf = PyUnicode_AS_UNICODE(__pyx_v_s);
#44 149.4             |                 ^~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
#44 149.4         580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
#44 149.4             |                                             ^~~~~~~~~~~~~~~~~~~
#44 149.4       aarch64-linux-gnu-g++ -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 build/temp.linux-aarch64-cpython-39/src/openstep_plist/parser.o -L/usr/lib -o build/lib.linux-aarch64-cpython-39/openstep_plist/parser.cpython-39-aarch64-linux-gnu.so
#44 149.4       building 'openstep_plist.util' extension
#44 149.4       aarch64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -ffile-prefix-map=/build/python3.9-PN012d/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DCYTHON_TRACE_NOGIL=1 -I/install/include -I/usr/include/python3.9 -c src/openstep_plist/util.cpp -o build/temp.linux-aarch64-cpython-39/src/openstep_plist/util.o -std=c++11
#44 149.4       aarch64-linux-gnu-g++ -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 build/temp.linux-aarch64-cpython-39/src/openstep_plist/util.o -L/usr/lib -o build/lib.linux-aarch64-cpython-39/openstep_plist/util.cpython-39-aarch64-linux-gnu.so
#44 149.4       building 'openstep_plist.writer' extension
#44 149.4       aarch64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -ffile-prefix-map=/build/python3.9-PN012d/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DCYTHON_TRACE_NOGIL=1 -I/install/include -I/usr/include/python3.9 -c src/openstep_plist/writer.cpp -o build/temp.linux-aarch64-cpython-39/src/openstep_plist/writer.o -std=c++11
#44 149.4       In file included from /usr/include/c++/10/bits/stl_algobase.h:67,
#44 149.4                        from /usr/include/c++/10/bits/char_traits.h:39,
#44 149.4                        from /usr/include/c++/10/ios:40,
#44 149.4                        from src/openstep_plist/writer.cpp:1186:
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h: In instantiation of ‘class std::back_insert_iterator<std::vector<wchar_t>&>’:
#44 149.4       src/openstep_plist/writer.cpp:4464:201:   required from here
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h:600:19: error: forming pointer to reference type ‘std::vector<wchar_t>&’
#44 149.4         600 |       _Container* container;
#44 149.4             |                   ^~~~~~~~~
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h:637:7: error: ‘std::vector<wchar_t>&’ is not a class, struct, or union type
#44 149.4         637 |       operator=(const typename _Container::value_type& __value)
#44 149.4             |       ^~~~~~~~
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h:645:7: error: ‘std::vector<wchar_t>&’ is not a class, struct, or union type
#44 149.4         645 |       operator=(typename _Container::value_type&& __value)
#44 149.4             |       ^~~~~~~~
#44 149.4       src/openstep_plist/writer.cpp: In function ‘PyObject* __pyx_f_14openstep_plist_6writer_6Writer__getvalue(__pyx_obj_14openstep_plist_6writer_Writer*)’:
#44 149.4       src/openstep_plist/writer.cpp:4530:113: warning: ‘PyObject* PyUnicode_FromUnicode(const Py_UNICODE*, Py_ssize_t)’ is deprecated [-Wdeprecated-declarations]
#44 149.4        4530 |   __pyx_t_1 = PyUnicode_FromUnicode(((Py_UNICODE const *)__pyx_v_self->dest->data()), __pyx_v_self->dest->size()); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 170, __pyx_L1_error)
#44 149.4             |                                                                                                                 ^
#44 149.4       In file included from /usr/include/python3.9/unicodeobject.h:1026,
#44 149.4                        from /usr/include/python3.9/Python.h:97,
#44 149.4                        from src/openstep_plist/writer.cpp:39:
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:551:42: note: declared here
#44 149.4         551 | Py_DEPRECATED(3.3) PyAPI_FUNC(PyObject*) PyUnicode_FromUnicode(
#44 149.4             |                                          ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/writer.cpp: In function ‘Py_ssize_t __pyx_f_14openstep_plist_6writer_6Writer_write_unquoted_string(__pyx_obj_14openstep_plist_6writer_Writer*, PyObject*)’:
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:279:45: warning: ‘Py_UNICODE* PyUnicode_AsUnicode(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         279 |       PyUnicode_AsUnicode(_PyObject_CAST(op)))
#44 149.4             |                                             ^
#44 149.4       src/openstep_plist/writer.cpp:5937:15: note: in expansion of macro ‘PyUnicode_AS_UNICODE’
#44 149.4        5937 |   __pyx_v_s = PyUnicode_AS_UNICODE(__pyx_v_string);
#44 149.4             |               ^~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
#44 149.4         580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
#44 149.4             |                                             ^~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:451:75: warning: ‘Py_ssize_t _PyUnicode_get_wstr_length(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         451 | #define PyUnicode_WSTR_LENGTH(op) _PyUnicode_get_wstr_length((PyObject*)op)
#44 149.4             |                                                                           ^
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:261:7: note: in expansion of macro ‘PyUnicode_WSTR_LENGTH’
#44 149.4         261 |       PyUnicode_WSTR_LENGTH(op) :                    \
#44 149.4             |       ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/writer.cpp:5946:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        5946 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_string);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here
#44 149.4         446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {
#44 149.4             |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:262:52: warning: ‘Py_UNICODE* PyUnicode_AsUnicode(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         262 |       ((void)PyUnicode_AsUnicode(_PyObject_CAST(op)),\
#44 149.4             |                                                    ^
#44 149.4       src/openstep_plist/writer.cpp:5946:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        5946 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_string);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
#44 149.4         580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
#44 149.4             |                                             ^~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:451:75: warning: ‘Py_ssize_t _PyUnicode_get_wstr_length(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         451 | #define PyUnicode_WSTR_LENGTH(op) _PyUnicode_get_wstr_length((PyObject*)op)
#44 149.4             |                                                                           ^
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:264:8: note: in expansion of macro ‘PyUnicode_WSTR_LENGTH’
#44 149.4         264 |        PyUnicode_WSTR_LENGTH(op)))
#44 149.4             |        ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/writer.cpp:5946:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        5946 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_string);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here
#44 149.4         446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {
#44 149.4             |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/writer.cpp: In function ‘Py_ssize_t __pyx_f_14openstep_plist_6writer_6Writer_write_string(__pyx_obj_14openstep_plist_6writer_Writer*, PyObject*)’:
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:279:45: warning: ‘Py_UNICODE* PyUnicode_AsUnicode(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         279 |       PyUnicode_AsUnicode(_PyObject_CAST(op)))
#44 149.4             |                                             ^
#44 149.4       src/openstep_plist/writer.cpp:6008:15: note: in expansion of macro ‘PyUnicode_AS_UNICODE’
#44 149.4        6008 |   __pyx_v_s = PyUnicode_AS_UNICODE(__pyx_v_string);
#44 149.4             |               ^~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
#44 149.4         580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
#44 149.4             |                                             ^~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:451:75: warning: ‘Py_ssize_t _PyUnicode_get_wstr_length(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         451 | #define PyUnicode_WSTR_LENGTH(op) _PyUnicode_get_wstr_length((PyObject*)op)
#44 149.4             |                                                                           ^
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:261:7: note: in expansion of macro ‘PyUnicode_WSTR_LENGTH’
#44 149.4         261 |       PyUnicode_WSTR_LENGTH(op) :                    \
#44 149.4             |       ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/writer.cpp:6017:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        6017 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_string);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here
#44 149.4         446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {
#44 149.4             |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:262:52: warning: ‘Py_UNICODE* PyUnicode_AsUnicode(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         262 |       ((void)PyUnicode_AsUnicode(_PyObject_CAST(op)),\
#44 149.4             |                                                    ^
#44 149.4       src/openstep_plist/writer.cpp:6017:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        6017 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_string);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
#44 149.4         580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
#44 149.4             |                                             ^~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:451:75: warning: ‘Py_ssize_t _PyUnicode_get_wstr_length(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         451 | #define PyUnicode_WSTR_LENGTH(op) _PyUnicode_get_wstr_length((PyObject*)op)
#44 149.4             |                                                                           ^
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:264:8: note: in expansion of macro ‘PyUnicode_WSTR_LENGTH’
#44 149.4         264 |        PyUnicode_WSTR_LENGTH(op)))
#44 149.4             |        ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/writer.cpp:6017:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        6017 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_string);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here
#44 149.4         446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {
#44 149.4             |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/writer.cpp: In function ‘Py_ssize_t __pyx_f_14openstep_plist_6writer_6Writer_write_short_float_repr(__pyx_obj_14openstep_plist_6writer_Writer*, PyObject*)’:
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:279:45: warning: ‘Py_UNICODE* PyUnicode_AsUnicode(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         279 |       PyUnicode_AsUnicode(_PyObject_CAST(op)))
#44 149.4             |                                             ^
#44 149.4       src/openstep_plist/writer.cpp:6150:15: note: in expansion of macro ‘PyUnicode_AS_UNICODE’
#44 149.4        6150 |   __pyx_v_s = PyUnicode_AS_UNICODE(__pyx_v_string);
#44 149.4             |               ^~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
#44 149.4         580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
#44 149.4             |                                             ^~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:451:75: warning: ‘Py_ssize_t _PyUnicode_get_wstr_length(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         451 | #define PyUnicode_WSTR_LENGTH(op) _PyUnicode_get_wstr_length((PyObject*)op)
#44 149.4             |                                                                           ^
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:261:7: note: in expansion of macro ‘PyUnicode_WSTR_LENGTH’
#44 149.4         261 |       PyUnicode_WSTR_LENGTH(op) :                    \
#44 149.4             |       ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/writer.cpp:6159:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        6159 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_string);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here
#44 149.4         446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {
#44 149.4             |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:262:52: warning: ‘Py_UNICODE* PyUnicode_AsUnicode(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         262 |       ((void)PyUnicode_AsUnicode(_PyObject_CAST(op)),\
#44 149.4             |                                                    ^
#44 149.4       src/openstep_plist/writer.cpp:6159:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        6159 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_string);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:580:45: note: declared here
#44 149.4         580 | Py_DEPRECATED(3.3) PyAPI_FUNC(Py_UNICODE *) PyUnicode_AsUnicode(
#44 149.4             |                                             ^~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:451:75: warning: ‘Py_ssize_t _PyUnicode_get_wstr_length(PyObject*)’ is deprecated [-Wdeprecated-declarations]
#44 149.4         451 | #define PyUnicode_WSTR_LENGTH(op) _PyUnicode_get_wstr_length((PyObject*)op)
#44 149.4             |                                                                           ^
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:264:8: note: in expansion of macro ‘PyUnicode_WSTR_LENGTH’
#44 149.4         264 |        PyUnicode_WSTR_LENGTH(op)))
#44 149.4             |        ^~~~~~~~~~~~~~~~~~~~~
#44 149.4       src/openstep_plist/writer.cpp:6159:20: note: in expansion of macro ‘PyUnicode_GET_SIZE’
#44 149.4        6159 |   __pyx_v_length = PyUnicode_GET_SIZE(__pyx_v_string);
#44 149.4             |                    ^~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/python3.9/cpython/unicodeobject.h:446:26: note: declared here
#44 149.4         446 | static inline Py_ssize_t _PyUnicode_get_wstr_length(PyObject *op) {
#44 149.4             |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
#44 149.4       In file included from /usr/include/c++/10/bits/stl_algobase.h:67,
#44 149.4                        from /usr/include/c++/10/bits/char_traits.h:39,
#44 149.4                        from /usr/include/c++/10/ios:40,
#44 149.4                        from src/openstep_plist/writer.cpp:1186:
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h: In instantiation of ‘std::back_insert_iterator<_Container>::back_insert_iterator(_Container&) [with _Container = std::vector<wchar_t>&]’:
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h:686:14:   required from ‘std::back_insert_iterator<_Container> std::back_inserter(_Container&) [with _Container = std::vector<wchar_t>&]’
#44 149.4       src/openstep_plist/writer.cpp:4464:201:   required from here
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h:614:40: error: using invalid field ‘std::back_insert_iterator<_Container>::container’
#44 149.4         614 |       : container(std::__addressof(__x)) { }
#44 149.4             |                                        ^
#44 149.4       In file included from /usr/include/c++/10/bits/char_traits.h:39,
#44 149.4                        from /usr/include/c++/10/ios:40,
#44 149.4                        from src/openstep_plist/writer.cpp:1186:
#44 149.4       /usr/include/c++/10/bits/stl_algobase.h: In instantiation of ‘static _OI std::__copy_move<false, false, std::random_access_iterator_tag>::__copy_m(_II, _II, _OI) [with _II = const wchar_t*; _OI = std::back_insert_iterator<std::vector<wchar_t>&>]’:
#44 149.4       /usr/include/c++/10/bits/stl_algobase.h:472:30:   required from ‘_OI std::__copy_move_a2(_II, _II, _OI) [with bool _IsMove = false; _II = const wchar_t*; _OI = std::back_insert_iterator<std::vector<wchar_t>&>]’
#44 149.4       /usr/include/c++/10/bits/stl_algobase.h:506:42:   required from ‘_OI std::__copy_move_a1(_II, _II, _OI) [with bool _IsMove = false; _II = const wchar_t*; _OI = std::back_insert_iterator<std::vector<wchar_t>&>]’
#44 149.4       /usr/include/c++/10/bits/stl_algobase.h:514:31:   required from ‘_OI std::__copy_move_a(_II, _II, _OI) [with bool _IsMove = false; _II = const wchar_t*; _OI = std::back_insert_iterator<std::vector<wchar_t>&>]’
#44 149.4       /usr/include/c++/10/bits/stl_algobase.h:569:7:   required from ‘_OI std::copy(_II, _II, _OI) [with _II = const wchar_t*; _OI = std::back_insert_iterator<std::vector<wchar_t>&>]’
#44 149.4       src/openstep_plist/writer.cpp:4464:202:   required from here
#44 149.4       /usr/include/c++/10/bits/stl_algobase.h:380:18: error: no match for ‘operator=’ (operand types are ‘std::back_insert_iterator<std::vector<wchar_t>&>’ and ‘const wchar_t’)
#44 149.4         380 |        *__result = *__first;
#44 149.4             |        ~~~~~~~~~~^~~~~~~~~~
#44 149.4       In file included from /usr/include/c++/10/bits/stl_algobase.h:67,
#44 149.4                        from /usr/include/c++/10/bits/char_traits.h:39,
#44 149.4                        from /usr/include/c++/10/ios:40,
#44 149.4                        from src/openstep_plist/writer.cpp:1186:
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h:596:11: note: candidate: ‘std::back_insert_iterator<std::vector<wchar_t>&>& std::back_insert_iterator<std::vector<wchar_t>&>::operator=(const std::back_insert_iterator<std::vector<wchar_t>&>&)’
#44 149.4         596 |     class back_insert_iterator
#44 149.4             |           ^~~~~~~~~~~~~~~~~~~~
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h:596:11: note:   no known conversion for argument 1 from ‘const wchar_t’ to ‘const std::back_insert_iterator<std::vector<wchar_t>&>&’
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h:596:11: note: candidate: ‘std::back_insert_iterator<std::vector<wchar_t>&>& std::back_insert_iterator<std::vector<wchar_t>&>::operator=(std::back_insert_iterator<std::vector<wchar_t>&>&&)’
#44 149.4       /usr/include/c++/10/bits/stl_iterator.h:596:11: note:   no known conversion for argument 1 from ‘const wchar_t’ to ‘std::back_insert_iterator<std::vector<wchar_t>&>&&’
#44 149.4       error: command '/usr/bin/aarch64-linux-gnu-gcc' failed with exit code 1
#44 149.4       [end of output]
#44 149.4   
#44 149.4   note: This error originates from a subprocess, and is likely not a problem with pip.
#44 149.4   ERROR: Failed building wheel for openstep-plist
```

--- a/.github/workflows/wheels.yml
+++ b/.github/workflows/wheels.yml
@@ -19,8 +19,14 @@ jobs:
     env:
       CIBW_ARCHS: auto64
       CIBW_ARCHS_MACOS: "x86_64 universal2"
-      # skip 3.6 on all platforms; skip pypy for windows
-      CIBW_SKIP: cp36-* pp*-win_amd64
+      # Skip
+      #
+      # * Python 3.6 on all platforms,
+      # * Python 3.12 on all platforms, and
+      # * PyPy on Windows.
+      #
+      # TODO: Activate Python 3.12 when issue 18 is resolved.
+      CIBW_SKIP: cp36-* cp312-* pp*-win_amd64
     steps:
       - uses: actions/checkout@v2
         with:

--- a/src/openstep_plist/writer.pyx
+++ b/src/openstep_plist/writer.pyx
@@ -11,8 +11,6 @@ from cpython.unicode cimport (
 )
 from cpython.bytes cimport PyBytes_GET_SIZE
 from cpython.object cimport Py_SIZE
-from libcpp.algorithm cimport copy
-from libcpp.iterator cimport back_inserter
 from libcpp.vector cimport vector
 from libc.stdint cimport uint16_t
 cimport cython
@@ -163,7 +161,7 @@ cdef class Writer:
         self, const Py_UNICODE *s, Py_ssize_t length
     ) except +:
         self.dest.reserve(self.dest.size() + length)
-        copy(s, s + length, back_inserter(self.dest[0]))
+        self.dest.insert(self.dest.end(), s, s + length)
         return length
 
     cdef inline unicode _getvalue(self):

